src/compiler.js

```diff
@@ -38,16 +40,26 @@ module.exports.run = ({ prefix, favicons: options, logo, cache: cacheDirectory }
 
       delete compilation.assets[output];
 
-      return compilation.hooks.webappWebpackPluginBeforeEmit.promise(result).then(({ html, assets }) => {
-        for (const { name, contents } of result.assets) {
+      if(compilation.hooks) {
```

Comment:
Unfortunately this will only work if we are running on Webpack v4, which means we need some extra gymnastics to remain compatible to Webpack v3.

In [`compat.js`](https://github.com/brunocodutra/webapp-webpack-plugin/blob/master/src/compat.js) we already define a couple helpers to make it easier to support both versions, all we need to do is to add a new one to trigger our own hooks.

Something like the following should work (I haven't tested though):

```js
/* istanbul ignore next */
module.exports.waterfall = (tappable, hook, arg, callback) => {
  if (tappable.hooks) { /* Webpack >= 4.0 */
    tappable.hooks[camelCase(hook)].callAsync(arg, callback);
  } else {
    tappable.applyPluginsAsyncWaterfall(hook, arg, callback);
  }
};
```

It should then be used like this:

```js
waterfall(compilation, 'webapp-webpack-plugin-before-emit', result, ({ html, assets }) => {
  for (const { name, contents } of result.assets) {
    compilation.assets[name] = {
      source: () => contents,
      size: () => contents.length,
    };
  }

  return resolve(result.html);
});
```

Notice we use `hooks-named-like-this` in Webpack v3, whereas in Webpack v4 we use camel case.

Reply:
Also, since the helper already makes sure it works in both versions of Webpack 4, we won't need this conditional here.